salt/_modules/metalk8s_drain.py

```diff
@@ -0,0 +1,512 @@
+
+        def message(errors):
+            msgs = []
+            for key, msg in errors.items():
+                msgs.append("{0}: {1}".format(key, ", ".join(msg)))
+            return "; ".join(msgs)
+
+        if failures:
+            raise DrainException(message(failures))
+        if warnings:
+            log.warning("WARNING: %s", message(warnings))
+        return pods
+
+    def run_drain(self, dry_run=False, **kwargs):
+        '''Cordon the node if necessary, then drain it.'''
+        node_unschedulable = __salt__['metalk8s_kubernetes.node_unschedulable'](
+            self.node_name, **kwargs
+        )
+        if not node_unschedulable:
+            __salt__['metalk8s_kubernetes.node_cordon'](self.node_name, **kwargs)
+        return self.evict_pods_orchestrate(dry_run=dry_run)
+
+    def evict_pods_orchestrate(self, dry_run=False):
+        '''Delete or evict pods based on eviction support.
+
+        In practice, all the versions we support have eviction, so we just
+        evict the pods.
+        '''
+        try:
+            pods = self.get_pods_for_eviction()
+        except DrainException as exc:
+            if dry_run:
+                pods = None
+            raise CommandExecutionError(
+                (
+                    "The following are not deletable: {0}. "
+                    "You can ignore DaemonSet pods with the "
+                    "ignore_daemonset flag."
+                ).format(
+                    exc.message
+                )
+            )
+
+        if dry_run:
+            return "Prepared for deletion of pods: {0}".format(
+                ", ".join(pods) if pods else "no pods to delete."
+            )
+
+        def get_pod(namespace, name):
+            '''Retrieve pod information'''
+            try:
+                api_instance = kubernetes.client.CoreV1Api()
+                api_response = api_instance.list_namespaced_pod(
+                    namespace=namespace,
+                    field_selector='metadata.name={0}'.format(name)
+                )
+                return api_response
+            except (ApiException, HTTPError) as exc:
+                if isinstance(exc, ApiException) and exc.status == 404:
+                    return None
+
+                log.exception(
+                    'Exception when calling '
+                    'CoreV1Api->list_namespaced_daemon_set')
+                raise CommandExecutionError(exc)
+
+        try:
+            self.evict_pods(pods, get_pod)
+        except DrainTimeoutException as exc:
+            raise CommandExecutionError(
+                "Drain did not complete within {0}".format(exc.message)
+            )
+        except DrainException as exc:
+            remaining_pods = self.get_pods_for_eviction()
+            raise CommandExecutionError(
+                "Pending pods remaining when the error occurred: {0}".format(
+                    exc),
+                [pod.metadata.name for pod in remaining_pods]
```

Comment:
There was no correct documentation on this, so I went to look at the source for my own education.

```python
    def __init__(self, message='', info=None):
        [...]
        self.info = info
        if self.info:
            if exc_str_prefix:
                if exc_str_prefix[-1] not in '.?!':
                    exc_str_prefix += '.'
                exc_str_prefix += ' '

            [...]

            # For states, if self.info is a dict also provide an attribute
            # containing a nested output of the info dict without the changes
            # (since they will be in the 'changes' key of the state return and
            # this information would be redundant).
            if isinstance(self.info, dict):
                info_without_changes = copy.deepcopy(self.info)
                info_without_changes.pop('changes', None)
                if info_without_changes:
                    self.strerror_without_changes = \
                        exc_str_prefix + _nested_output(info_without_changes)
                [...
```

So it's OK to pass random stuff to it with the expectation that it'll be represented as usual. But that's not what your comment is about! And indeed this is redundant, especially given that currently there is no `DrainException` to catch so all this is useless.

Again, I need to resync with the `kubectl` code control flow. :angry: